### PR TITLE
ARROW-16809: [C++] Add Benchmarks for FilterNode

### PR DESCRIPTION
**Pull request by iChauster**
Add `filter_benchmark.cc` and add to `arrow/compute/exec/CMakeLists`

### CODE REVIEW

File: cpp/src/arrow/compute/exec/filter_benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      static_cast<double>(state.iterations() * num_batches), benchmark::Counter::kIsRate);
+}
+
+arrow::compute::Expression complex_expression =
+    less(less(field_ref("i64"), literal(20)), greater(field_ref("i64"), literal(0)));
+arrow::compute::Expression simple_expression =
+    less(call("negate", {field_ref("i64")}), literal(0));
+arrow::compute::Expression zero_copy_expression = is_valid((call(
+    "cast", {field_ref("i64")}, compute::CastOptions::Safe(timestamp(TimeUnit::NANO)))));
+arrow::compute::Expression ref_only_expression = less(field_ref("i64"), literal(0));
```
Comment (iChauster) [inline cpp/src/arrow/compute/exec/filter_benchmark.cc:150]: I'm not sure how we want to express the complexity of filter expressions, so I used a similar approach to `ExpressionOverhead` / `ProjectBenchmarks`. `zero_copy` seems to perform a bit better than the others because it uses `is_valid` (I cannot do the less operation on timestamps). 

Is there something I can change to make these more consistent?

**Comment by iChauster** [comment]
Hi @westonpace, this is a similar PR to #13314, just switched out the `ProjectNode` for a `FilterNode`. 

I had some questions about the expression complexity and if we would want to adjust that for these benchmarks!

**Review by westonpace — CHANGES_REQUESTED** [review]
Sorry for the delay in review.  This is a good start.  For a filtering operation I think there is an extra parameter which is the selectivity (what percentage of rows are kept).  I think it would be valuable to add that as a parameter but it would make test data generation more complicated.

Another interesting question, but maybe not one we have to answer here, is if it is faster to say...

`FILTER(and(lt(x, 100), gt(x, 0)))` or `FILTER(FILTER(lt(x, 100)), gt(x, 0))`.  In other words, is it faster to compute two N sized boolean arrays and then do one filtering pass or if it is faster to do two filtering passes, the first against an N sized boolean array and the second against the smaller array.  This benchmark would depend on the selectivity of the two filters I think.

**Comment by iChauster** [comment]
> For a filtering operation I think there is an extra parameter which is the selectivity (what percentage of rows are kept). I think it would be valuable to add that as a parameter but it would make test data generation more complicated.

@westonpace one idea I had to benchmark selectivity is perhaps using the existing 'null_percent/proportion' generators we already have, and then using `is_null` as the filter. Let me know if you think that would be the right approach.

File: cpp/src/arrow/compute/exec/filter_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    ASSERT_OK_AND_ASSIGN(std::shared_ptr<arrow::compute::ExecPlan> plan,
+                         ExecPlan::Make(&ctx));
+    // Source and sink nodes have no effect on the benchmark.
+    // Used for dummy purposes as they are referenced in InputReceived and InputFinished.
+    ASSERT_OK_AND_ASSIGN(
+        arrow::compute::ExecNode * source_node,
+        MakeExecNode("source", plan.get(), {},
+                     SourceNodeOptions{data.schema, data.gen(/*parallel=*/true,
+                                                             /*slow=*/false)}));
+    ASSERT_OK_AND_ASSIGN(arrow::compute::ExecNode * project_node,
```
[the remark this answers is not in the document]
Comment (iChauster) [reply cpp/src/arrow/compute/exec/filter_benchmark.cc]: Moved most of the logic into `benchmark_util.h`, reflected in `project_benchmark.cc` as well.

**Comment by westonpace** [comment]
Sorry it took so long.  I wanted to be able to run on my dev machine so was waiting for that to finish moving (had to ship it).  Thanks for the cleanup, it's very promising that we can share so much between nodes.  I think you've captured a good set of interesting cases here.

I agree with your conclusions.  We also seem to continue to have peak performance in the 10k-100k rows-per-batch range.

**Comment by westonpace** [comment]
Just a few minor nits.  This looks pretty good.

**Comment by iChauster** [comment]
Hi @westonpace, was wondering if you could publish the changes you requested -- I don't think I can see them yet.

**Comment by westonpace** [comment]
> Hi @westonpace, was wondering if you could publish the changes you requested -- I don't think I can see them yet.

@iChauster Done, sorry :facepalm:

**Comment by iChauster** [comment]
Excellent, if there's nothing else I think this is ready to go @westonpace!

File: cpp/src/arrow/util/benchmark_util.h [hunk 3]
```diff
@@ -15,10 +15,24 @@
 // specific language governing permissions and limitations
 // under the License.
 
+#include <condition_variable>
+#include <mutex>
```
Comment (westonpace) [inline cpp/src/arrow/util/benchmark_util.h:19]: Ah, this is going to be a problem.  We can't include `mutex` in public headers due to a Windows compatibility issue.  Can you create a `benchmark_util.cc` and move the definition of these methods there?

Comment (iChauster) [reply cpp/src/arrow/util/benchmark_util.h]: @westonpace , sorry if this is trivial, but I'm having some trouble here.

I committed some changes, but it seems like the compiler is having trouble finding third party dependencies imported in `benchmark_util.h` (like benchmarks, gtest) after including `benchmark_util.cc` in the `src/arrow/CMakeLists.txt`. Let me know if I'm missing a step.

Comment (iChauster) [reply cpp/src/arrow/util/benchmark_util.h]: Hi @westonpace, have you had a chance to take a look at this?

File: cpp/src/arrow/CMakeLists.txt [hunk 4]
```diff
@@ -193,6 +193,7 @@ set(ARROW_SRCS
     io/transform.cc
     util/async_util.cc
     util/basic_decimal.cc
+    # util/benchmark_util.cc
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/CMakeLists.txt:196]: Ah, nevermind, I think I see what you mean.

**Comment by westonpace** [comment]
How about this: https://github.com/apache/arrow/commit/f7109966366f389ba3bace571aa55571f70b2aec

* moves the benchmark utilities for exec plans into a new benchmark_util that is in the compute/exec folder.  This change isn't strictly necessary but helps keep the code separated.
* removes any dependency on googletest and instead changes the functions to return a status.  Then only the benchmark needs to depend on googletest (and can do ASSERT_OK in one spot).
* changes the build slightly so that the benchmark_util.cc file is only built when building benchmarks so that this code doesn't get included with the libarrow.so (or libarrow_compute.so) files.

Locally, tests seem to pass on my fork.

**Comment by iChauster** [comment]
@westonpace, that looks great, thanks for the help! I can run all the benchmarks on my system as well -- I think this is ready to go.

File: cpp/src/arrow/util/benchmark_util.h [hunk 5]
```diff
@@ -19,9 +19,8 @@
 #include <cstdint>
```
Comment (westonpace) [inline cpp/src/arrow/util/benchmark_util.h:19]: Can you revert the changes to this file?

* We like to keep 3rd party includes in their own group and not clustered with arrow includes.
* Closing namespaces don't need the `;`

**Review by westonpace — APPROVED** [review]
Approved once we revert the changes to `cpp/src/arrow/util/benchmark_util.h`

File: cpp/src/arrow/util/benchmark_util.h (hunk 5 above)
Comment (iChauster) [reply cpp/src/arrow/util/benchmark_util.h]: Sorry about that! Should be good here.

**Comment by westonpace** [comment]
The build failure appears unrelated.  I will merge.  Thanks again.
